### PR TITLE
simple fiat shamir transform

### PR DESCRIPTION
**Pull request by hduoc2003**

### CODE REVIEW

File: fri/src/fiat_shamir/mod.rs [hunk 1]
```diff
@@ -1,30 +1,68 @@
...
         Self {
-            data: Vec::new()
+            last_r: F::ZERO,
+            x: F::from(StdRng::from_entropy().gen::<u128>()),
+            i: 0,
+            gi: F::ZERO
         }
     }
 
     pub fn append(&mut self, merkle_root: &F) {
```
Comment (zk-steve) [inline fri/src/fiat_shamir/mod.rs]:
```suggestion
    pub fn append(&mut self, merkle_root: F) {
```

Comment (zk-steve) [reply fri/src/fiat_shamir/mod.rs]: Let the consumer of this method clone the merkle_root themself

File: fri/src/fiat_shamir/mod.rs [hunk 2]
```diff
@@ -1,30 +1,68 @@
 use ark_ff::PrimeField;
+use rand::{rngs::StdRng, Rng, SeedableRng};
+
+use crate::hasher::CustomizedHash;
 
 // A transcript is a list of Merkle root.
 pub struct Transcript<F: PrimeField>{
-    data: Vec<F>
+    last_r: F,
```
Comment (zk-steve) [inline fri/src/fiat_shamir/mod.rs:11]: write doc explaining the meaning of each vars

File: fri/src/fiat_shamir/mod.rs [hunk 3]
```diff
@@ -1,30 +1,68 @@
...
+mod tests {
+    use crate::fields::goldilocks::Fq;
+
+    use super::Transcript;
+
+    #[test]
+    fn test_generate_a_challenge() {
+        let mut transcript = Transcript::<Fq>::new();
+        let g = transcript.generate_a_challenge();
+        println!("{:?}", g);
```
Comment (zk-steve) [inline fri/src/fiat_shamir/mod.rs]: this is the test ?

File: fri/src/fiat_shamir/mod.rs [hunk 4]
```diff
@@ -1,30 +1,68 @@
...
+    }
+
+    #[test]
+    fn test_generate_challenge_list() {
+        let mut transcript = Transcript::<Fq>::new();
+        let g = transcript.generate_challenge_list(5);
+        println!("{:?}", g);
```
Comment (zk-steve) [inline fri/src/fiat_shamir/mod.rs]: this is the test ?

File: fri/src/fiat_shamir/mod.rs [hunk 5]
```diff
@@ -1,30 +1,68 @@
...
+        self.last_r = challenge;
+        return challenge;
     }
     // generate challenges with data.
-    pub fn generate_challenge_list(&self, number: usize) -> Vec<F>{
-        let challenges = Vec::new();
-        // Todo: implement generate function.
+    pub fn generate_challenge_list(&mut self, number: usize) -> Vec<F>{
+        let mut challenges = Vec::new();
+        for _i in 0..number {
```
Comment (zk-steve) [inline fri/src/fiat_shamir/mod.rs]:
```suggestion
        for _ in 0..number {
```

File: fri/src/fiat_shamir/mod.rs [hunk 6]
```diff
@@ -1,30 +1,68 @@
...
-        self.data.push(merkle_root.clone());
+        self.i += 1;
+        self.gi = merkle_root.clone();
     }
 
-    pub fn generate_a_challenge(&self) -> F {
-        let challenge = F::ZERO;
-        // Todo: implement generate function.
-        challenge
+    pub fn generate_a_challenge(&mut self) -> F {
```
Comment (zk-steve) [inline fri/src/fiat_shamir/mod.rs:65]: doc

**Review by VanhGer — COMMENTED** [review]
Your fiat-shamir is wrong.
I have two instance of Transcript, if I append it with the same data, it output is different. That's not true for random oracle model.

Here is the test
```rust
fn test_generate_a_challenge2() {
        let mut transcript = Transcript::<Fq>::new();
        let query = Fq::from(928459);

        let mut transcript2 = Transcript::<Fq>::new();
        transcript.append(query);
        transcript2.append(query);
        let c1 = transcript.generate_a_challenge();
        // transcript.append(query);
        let c2 = transcript2.generate_a_challenge();
        println!("{:?} {:?}", c1.0, c2.0);
        assert_eq!(c1, c2);
    }
```

**Pull request merged**